src/ML-UF3/pair_uf3.cpp

```diff
@@ -0,0 +1,1313 @@
+      for (int j = 1; j < num_of_elements + 1; j++) {
+        for (int k = 1; k < num_of_elements + 1; k++) {
+          if (setflag_3b[i][j][k] != 1)
+            error->all(FLERR,"UF3: Not all 3-body UF potentials are set, \n\
+                    missing potential file for {}-{}-{} interaction", i, j, k);
+        }
+      }
+    }
+  }
+
+  for (int i = 1; i < num_of_elements + 1; i++) {
+    for (int j = i; j < num_of_elements + 1; j++) {
+      UFBS2b[i][j] = uf3_pair_bspline(lmp, n2b_knot[i][j], n2b_coeff[i][j],
+              knot_spacing_type_2b[i][j]);
+      UFBS2b[j][i] = UFBS2b[i][j];
+    }
+    if (pot_3b) {
+      for (int j = 1; j < num_of_elements + 1; j++) {
+        for (int k = j; k < num_of_elements + 1; k++) {
+          std::string key = std::to_string(i) + std::to_string(j) + std::to_string(k);
+          UFBS3b[i][j][k] =
+              uf3_triplet_bspline(lmp, n3b_knot_matrix[i][j][k], n3b_coeff_matrix[key],
+                      knot_spacing_type_3b[i][j][k]);
+          std::string key2 = std::to_string(i) + std::to_string(k) + std::to_string(j);
+          UFBS3b[i][k][j] =
+              uf3_triplet_bspline(lmp, n3b_knot_matrix[i][k][j], n3b_coeff_matrix[key2],
+                      knot_spacing_type_3b[i][k][j]);
+        }
+      }
+    }
+  }
+}
+
+void PairUF3::compute(int eflag, int vflag)
+{
+  int i, j, k, ii, jj, kk, inum, jnum, itype, jtype, ktype;
+  double xtmp, ytmp, ztmp, delx, dely, delz, evdwl, fpair, fx, fy, fz;
+  double del_rji[3], del_rki[3], del_rkj[3];
+  double fij[3], fik[3], fjk[3];
+  double fji[3], fki[3], fkj[3];
+  double Fi[3], Fj[3], Fk[3];
+  double rsq, rij, rik, rjk;
+  int *ilist, *jlist, *numneigh, **firstneigh;
+
+  ev_init(eflag, vflag);
+
+  double **x = atom->x;
+  double **f = atom->f;
+  int *type = atom->type;
+  int nlocal = atom->nlocal;
+  int newton_pair = force->newton_pair;
+
+  inum = list->inum;
+  ilist = list->ilist;
+  numneigh = list->numneigh;
+  firstneigh = list->firstneigh;
+  // loop over neighbors of my atoms
+  for (ii = 0; ii < inum; ii++) {
+    evdwl = 0;
+    i = ilist[ii];
+    xtmp = x[i][0];
+    ytmp = x[i][1];
+    ztmp = x[i][2];
+    itype = type[i];
+    jlist = firstneigh[i];
+    jnum = numneigh[i];
+    int numshort = 0;
+    for (jj = 0; jj < jnum; jj++) {
+      fx = 0;
+      fy = 0;
+      fz = 0;
+      j = jlist[jj];
+      j &= NEIGHMASK;
+
+      delx = xtmp - x[j][0];
+      dely = ytmp - x[j][1];
+      delz = ztmp - x[j][2];
+
+      rsq = delx * delx + dely * dely + delz * delz;
+      jtype = type[j];
+      if (rsq < cutsq[itype][jtype]) {
+        rij = sqrt(rsq);
+
+        if (pot_3b) {
+          if (rij <= cut_3b_list[itype][jtype]) {
+            neighshort[numshort] = j;
+            if (numshort >= maxshort - 1) {
+              maxshort += maxshort / 2;
+              memory->grow(neighshort, maxshort, "pair:neighshort");
+            }
+            numshort = numshort + 1;
+          }
+        }
```

Comment:
I agree storing an array of rij will reduce the number of computations later on.
But I also need to know the type of atoms in the 3b interaction evaluation. So I need to keep the current `neighshort`.

Also, in practice, the 3b cutoffs are always shorter than 2-body cutoffs (usually by a couple of angstroms). Do you still think I need to make the `rij_list` of the size of `inum`?

Reply:
> Types are straightforward to track: if `jj` runs from 0 to `jnum`, the distance `rij_list[jj]` belongs to particle `jlist[jj]` which has particle `type type[jlist[jj]]`.

`jnum` is decided by the cutoff of 3body interaction and, in practice, is always less than 2body.

In `neighshort`, I am storing `jlist[jj]` during the 2body loop. Not every `j` in 2body interactions forms a valid 3body triangle. 

So, if my understanding of your suggestion is correct, you are suggesting to set `jnum` of 3body loop equal to `jnum` of 3body loop(?). If that is done, a lot of iterations in the 3body loop will do nothing because of the distance  cutoff checks. 

I apologize if I am not understanding your suggestion correctly.